text/0016-revise_transform_processing.md

```diff
@@ -0,0 +1,192 @@
+
+Let's have more support for the basic idea that functions that have different signatures _do different things_. That isn't really honored here. It's not clear from the above example that download and transform have any predeclared order or capabilities. Transform functions as implemented appear to differ in name only, and the predeclared call order runs the risk of becoming arbitrary. 
+
+To solve both of these problems, I'm proposing the following changes:
+* replace the notion of a transform function with a set of predefined _special functions_, where each function has a required signature that fits it's needs
+* replace the "chain of tranform functions" with a single function that applies programmitic changes & returns the finalized dataset
+* all _special functions_ accept a passed `context` object for moving state across special function calls
+* remove the requirement that the final transform function return a dataset
+
+Here's a an example that re-writes the above transform script using the propsosed changes:
+
+```python
+load("http.sky", "http")
+
+def download(context):
+  res = http.get("https://api.example.com/cats.json")
+  return res.json()
+
+def transform(ds, context):
+  cats = context.download
+  ds.set_body(cats)
+  ds.set_meta("title", "a list of cats from example.com")
+```
+
+Here the notion of chaining transform functions is gone. Instead, there is now one point entry, a function, `transform` that accepts a dataset. `transform` acts as a "main" function that Qri calls, passing in the user-supplied dataset snapshot (whatever the user provided from either the command line or API). `transform` retains is previous sandbox qualities: access to the user's local qri repository, but no network access, and is now always provided with the user-input dataset.
+
+`download` is an example of a _special function_. Special functions are predefined function names and signatures the Qri environment knows to look for. Unlike our current transform functions, special functions have varying signatures, which must be conveyed through documentation. If a transform script defines a special function, Qri will call it and place the return value of the function at `context.[special_function_name]` before calling the main `transform` function. The main `transform` now also has access to the result of all _special function_ calls via the passed in `context`.
+
+`context` is designed to solve the issue of moving state around during a transform. All special functions will accept a `context` argument. The Qri environment will populate `context` with necessary transfrom state such as `transform.config` and `transform.secret` values. `context` also has an API for passing arbitrary user data via `context.set("key", Value)` and `context.get("key")`.
+
```

Comment:
Does this mean that the secrets and config will be located at `context.transform.secrets`, rather then using qri.get_secrets(key)?

Reply:
oooh v good question. I was thinking the new API would be `context.get_config` and `context.get_secret`, both as functions that take a name, and yes, both would move off of `qri` and onto `context`.

By making them function calls it'll be easier for us to tell which config and secret values a script is making use of so we can later tell other users of the script stuff like "this transform requires the following secrets: ..."